File: app/src/main/java/com/example/android/tictactoe/utils/TicTacToeUtils.java
package com.example.android.tictactoe.utils;

import android.widget.Button;

import com.example.android.tictactoe.R;

public class TicTacToeUtils {

    public static final int SINGLE_PLAYER_EASY_MODE = 1;
    public static final int SINGLE_PLAYER_MEDIUM_MODE = 2;
    public static final int SINGLE_PLAYER_IMPOSSIBLE_MODE = 3;
    public static final int TWO_PLAYER_MODE = 4;
    public static final int PLAYER_X_PLAYED_VALUE = 1;
    public static final int PLAYER_O_PLAYED_VALUE = 4;
    public static final int NON_PLAYED_VALUE = 0;

    private TicTacToeUtils() { /* cannot be instantiated */ }

    /**
     * Returns true if game is played in single player mode.
     *
     * @param gameMode the game mode selected by the user
     */
    public static boolean isSinglePlayerMode(int gameMode) {
        if (gameMode == SINGLE_PLAYER_EASY_MODE
                || gameMode == SINGLE_PLAYER_MEDIUM_MODE
                || gameMode == SINGLE_PLAYER_IMPOSSIBLE_MODE) {
            return true;
        } else return false;
    }

    /**
     * Returns an array of values from the two dimensional array board.
     * Conversion is done because savedInstantState can not save two dimensional arrays.
     *
     * @param twoDimBoard a two dimensional array to be converted.
     * @param boardSize   the size of the board
     * @return oneDimBoard an array of integers
     */
    public static int[] convertBoardToOneDim(int boardSize, int[][] twoDimBoard) {
        int[] oneDimBoard = new int[boardSize * boardSize];
        int count = 0;
        for (int i = 0; i < boardSize; i++) {
            for (int j = 0; j < boardSize; j++) {
                oneDimBoard[count] = twoDimBoard[i][j];
                count++;
            }
        }
        return oneDimBoard;
    }

    /**
     * Returns a two dimensional array board of values from the one dimensional array.
     *
     * @param oneDimBoard an array to be converted.
     * @param boardSize   the size of the board
     * @return twoDimBoard a two dimensional array of integers for the board
     */
    public static int[][] convertBoardToTwoDim(int boardSize, int[] oneDimBoard) {
        int[][] twoDimBoard = new int[boardSize][boardSize];
        int count = 0;
        for (int i = 0; i < boardSize; i++) {
            for (int j = 0; j < boardSize; j++) {
                twoDimBoard[i][j] = oneDimBoard[count];
                count++;
            }
        }
        return twoDimBoard;
    }

    public static int[][] initBoardWithZeros(int boardSize) {
        int[][] board = new int[boardSize][boardSize];
        for (int row = 0; row < boardSize; row++) {
            for (int column = 0; column < boardSize; column++) {
                // Fill the board with zeros(0)
                board[row][column] = 0;
            }
        }
        return board;
    }

    public static void setTextOnButtonPlayed(boolean PLAYER_X_TURN, Button row0col0) {
        if (PLAYER_X_TURN) {
            row0col0.setText(R.string.string_x);
        } else {
            row0col0.setText(R.string.string_o);
        }
    }

    public static void disableButton(Button row0col0) {
        row0col0.setEnabled(false);
    }

}

File: app/src/main/java/com/zizohanto/android/tictactoe/utils/TicTacToeUtils.kt
package com.zizohanto.android.tictactoe.utils

import android.widget.Button
import com.zizohanto.android.tictactoe.R

object TicTacToeUtils {
    const val SINGLE_PLAYER_EASY_MODE = 1
    const val SINGLE_PLAYER_MEDIUM_MODE = 2
    const val SINGLE_PLAYER_IMPOSSIBLE_MODE = 3
    const val TWO_PLAYER_MODE = 4
    const val PLAYER_X_PLAYED_VALUE = 1
    const val PLAYER_O_PLAYED_VALUE = 4
    const val NON_PLAYED_VALUE = 0

    /**
     * Returns true if game is played in single player mode.
     *
     * @param gameMode the game mode selected by the user
     */
    fun isSinglePlayerMode(gameMode: Int) =
            gameMode == SINGLE_PLAYER_EASY_MODE
                    || gameMode == SINGLE_PLAYER_MEDIUM_MODE
                    || gameMode == SINGLE_PLAYER_IMPOSSIBLE_MODE

    /**
     * Returns an array of values from the two dimensional array board.
     * Conversion is done because savedInstantState can not save two dimensional arrays.
     *
     * @param twoDimBoard a two dimensional array to be converted.
     * @param boardSize   the size of the board
     * @return oneDimBoard an array of integers
     */
    fun convertBoardToOneDim(boardSize: Int, twoDimBoard: Array<IntArray>): IntArray {
        val oneDimBoard = IntArray(boardSize * boardSize)
        var count = 0
        for (i in 0 until boardSize) {
            for (j in 0 until boardSize) {
                oneDimBoard[count] = twoDimBoard[i][j]
                count++
            }
        }
        return oneDimBoard
    }

    /**
     * Returns a two dimensional array board of values from the one dimensional array.
     *
     * @param oneDimBoard an array to be converted.
     * @param boardSize   the size of the board
     * @return twoDimBoard a two dimensional array of integers for the board
     */
    fun convertBoardToTwoDim(boardSize: Int, oneDimBoard: IntArray): Array<IntArray> {
        val twoDimBoard = Array(boardSize) { IntArray(boardSize) }
        var count = 0
        for (i in 0 until boardSize) {
            for (j in 0 until boardSize) {
                twoDimBoard[i][j] = oneDimBoard[count]
                count++
            }
        }
        return twoDimBoard
    }

    fun initBoardWithZeros(boardSize: Int): Array<IntArray> {
        val board = Array(boardSize) { IntArray(boardSize) }
        for (row in 0 until boardSize) {
            for (column in 0 until boardSize) {
                // Fill the board with zeros(0)
                board[row][column] = 0
            }
        }
        return board
    }

    fun setTextOnButtonPlayed(isPlayerXTurn: Boolean, row0col0: Button) {
        if (isPlayerXTurn) {
            row0col0.setText(R.string.string_x)
        } else {
            row0col0.setText(R.string.string_o)
        }
    }

    fun disableButton(row0col0: Button) {
        row0col0.isEnabled = false
    }
}